src/cargo/core/resolver/mod.rs

```diff
@@ -98,21 +98,93 @@ type ResolveResult = CargoResult<CargoResult<Box<Context>>>;
 type DepInfo = (Dependency, Vec<Rc<Summary>>, Vec<String>);
 
 impl Resolve {
-    fn new(root: PackageId) -> Resolve {
-        let mut g = Graph::new();
-        g.add(root.clone(), &[]);
-        Resolve { graph: g, root: root, features: HashMap::new(), metadata: None }
-    }
+    pub fn merge_from(&mut self, previous: &Resolve) -> CargoResult<()> {
+        // Given a previous instance of resolve, it should be forbidden to ever
+        // have a checksums which *differ*. If the same package id has differing
+        // checksums, then something has gone wrong such as:
+        //
+        // * Something got seriously corrupted
+        // * A "mirror" isn't actually a mirror as some changes were made
+        // * A replacement source wasn't actually a replacment, some changes
+        //   were made
+        //
+        // In all of these cases, we want to report an error to indicate that
+        // something is awry. Normal execution (esp just using crates.io) should
+        // never run into this.
+        for (id, cksum) in previous.checksums.iter() {
+            if let Some(mine) = self.checksums.get(id) {
+                if mine == cksum {
+                    continue
+                }
+
+                // If the previous checksum wasn't calculated, the current
+                // checksum is `Some`. This may indicate that a source was
+                // erroneously replaced or was replaced with something that
+                // desires stronger checksum guarantees than can be afforded
+                // elsewhere.
+                if cksum.is_none() {
+                    bail!("\
+checksum for `{}` was not previously calculated, but a checksum could now \
+be calculated
+
+this could be indicative of a few possible situations:
+
+    * the source `{}` did not previously support checksums,
+      but was replaced with one that does
+    * newer Cargo implementations know how to checksum this source, but this
+      older implementation does not
+    * the lock file is corrupt
+", id, id.source_id())
```

Comment:
This is actually a bit of an interesting error case in the sense that I'm not really sure what *can* be done here. If Cargo never changed then this should never be seen except in the case that something is corrupt or it's an internal error. This *may* be able to be fixed by upgrading the Cargo in use, but that's also not necessarily guaranteed to work.

Ah and for the messages below it's kinda the same thing as well, I'm not really sure what *can* be done. The most likely-to-be-seen one, mismatching checksums, will likely be attributed to something outside the user's control, e.g.:

* The replacement registry legitimately has a different crate
* The network interfered with the download at some point

In either of these situations there's unfortunately not really a lot that *can* be done :(